sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java

```diff
@@ -67,112 +72,169 @@
+  private static String tableName;
+
+  @Rule
+  public TestPipeline pipelineWrite = TestPipeline.create();
+  @Rule
+  public TestPipeline pipelineRead = TestPipeline.create();
 
   @BeforeClass
   public static void setup() throws SQLException {
     PipelineOptionsFactory.register(IOTestPipelineOptions.class);
     IOTestPipelineOptions options = TestPipeline.testingPipelineOptions()
         .as(IOTestPipelineOptions.class);
 
-    // We do dataSource set up in BeforeClass rather than Before since we don't need to create a new
-    // dataSource for each test.
     dataSource = JdbcTestDataSet.getDataSource(options);
+
+    tableName = JdbcTestDataSet.getWriteTableName();
+    JdbcTestDataSet.createDataTable(dataSource, tableName);
   }
 
   @AfterClass
   public static void tearDown() throws SQLException {
-    // Only do write table clean up once for the class since we don't want to clean up after both
-    // read and write tests, only want to do it once after all the tests are done.
-    JdbcTestDataSet.cleanUpDataTable(dataSource, writeTableName);
+    JdbcTestDataSet.cleanUpDataTable(dataSource, tableName);
   }
 
-  private static class CreateKVOfNameAndId implements JdbcIO.RowMapper<KV<String, Integer>> {
-    @Override
-    public KV<String, Integer> mapRow(ResultSet resultSet) throws Exception {
-      KV<String, Integer> kv =
-          KV.of(resultSet.getString("name"), resultSet.getInt("id"));
-      return kv;
-    }
+  /**
+   * Tests writing then reading data for a postgres database.
+   */
+  @Test
+  public void testWriteThenRead() {
+    runWrite();
+    runRead();
   }
 
-  private static class PutKeyInColumnOnePutValueInColumnTwo
-      implements JdbcIO.PreparedStatementSetter<KV<Integer, String>> {
-    @Override
-    public void setParameters(KV<Integer, String> element, PreparedStatement statement)
-                    throws SQLException {
-      statement.setInt(1, element.getKey());
-      statement.setString(2, element.getValue());
-    }
-  }
+  /**
+   * Writes the test dataset to postgres.
+   *
+   * <p>This test does not attempt to validate the data - we do so in the read test. This does make
+   * it harder to tell whether a test failed in the write or read phase, but the tests are much
+   * easier to maintain (don't need any separate code to write test data for read tests to
+   * the database.)
+   */
+  private void runWrite() {
+    pipelineWrite.apply(GenerateSequence.from(0).to(JdbcTestDataSet.EXPECTED_ROW_COUNT))
+        .apply(ParDo.of(new DeterministicallyConstructTestRowFn()))
+        .apply(JdbcIO.<TestRow>write()
+            .withDataSourceConfiguration(JdbcIO.DataSourceConfiguration.create(dataSource))
+            .withStatement(String.format("insert into %s values(?, ?)", tableName))
+            .withPreparedStatementSetter(new PrepareStatementFromTestRow()));
 
-  @Rule
-  public TestPipeline pipeline = TestPipeline.create();
+    pipelineWrite.run().waitUntilFinish();
```

Comment:
We don't validate the data that was written in this test because the read ITest is going to verify it immediately after the write test writes it. 

But I agree with your question about testing for scale. This test currently only runs for 1000 rows, but you'll note that that is captured in a few small variables. 

My thought is that we'll do something like what is captured in https://github.com/ssisk/beam/commit/7f96c844dd1b6181ace87303ac5574cf11cbd78f and have 2 or 3 different test sizes we run - one small (1000 rows), and one large (10/50 million rows) - the test will be parameterized and we'll be able to pass in a command line option for whether we want the test run to be small or large. 

What do you think about that? That will enable us to easily test at the large scale.